core/src/main/java/io/questdb/griffin/engine/join/HashOuterJoinFilteredLightRecordCursorFactory.java

```diff
@@ -0,0 +1,237 @@
+            this.isOpen = true;
+        }
+
+        @Override
+        public void close() {
+            if (isOpen) {
+                isOpen = false;
+                joinKeyMap.close();
+                slaveChain.close();
+                super.close();
+            }
+        }
+
+        @Override
+        public Record getRecord() {
+            return record;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (slaveChainCursor != null && slaveChainCursor.hasNext()) {
+                record.hasSlave(true);
+                while (slaveChainCursor.hasNext()) {
+                    slaveCursor.recordAt(slaveRecord, slaveChainCursor.next());
+                    if (filter.getBool(record)) {
+                        return true;
+                    } else {
+                        slaveChainCursor.next();
+                    }
+                }
+            }
+
+            if (masterCursor.hasNext()) {
+                MapKey key = joinKeyMap.withKey();
+                key.put(masterRecord, masterKeySink);
+                MapValue value = key.findValue();
+                if (value != null) {
+                    slaveChainCursor = slaveChain.getCursor(value.getLong(0));
+                    record.hasSlave(true);
+                    while (slaveChainCursor.hasNext()) {
+                        slaveCursor.recordAt(slaveRecord, slaveChainCursor.next());
+                        if (filter.getBool(record)) {
+                            return true;
+                        }
+                    }
+                }
+
+                slaveChainCursor = null;
+                record.hasSlave(false);
+                return true;
+            }
+
+            return false;
+        }
+
+        @Override
+        public long size() {
+            return -1L;
+        }
+
+        @Override
+        public void toTop() {
+            masterCursor.toTop();
+            slaveChainCursor = null;
+        }
+
+        private void buildMapOfSlaveRecords(RecordCursor slaveCursor, SqlExecutionCircuitBreaker circuitBreaker) {
+            try {
+                final Record record = slaveCursor.getRecord();
+                while (slaveCursor.hasNext()) {
+                    circuitBreaker.statefulThrowExceptionIfTripped();
+                    MapKey key = joinKeyMap.withKey();
+                    key.put(record, slaveKeySink);
+                    MapValue value = key.createValue();
+                    if (value.isNew()) {
+                        final long offset = slaveChain.put(record.getRowId(), -1);
+                        value.putLong(0, offset);
+                        value.putLong(1, offset);
+                    } else {
+                        value.putLong(1, slaveChain.put(record.getRowId(), value.getLong(1)));
+                    }
+                }
+            } catch (Throwable t) {
+                close();
```

Comment:
I'd move this close() to the `public RecordCursor getCursor(SqlExecutionContext executionContext)`, this is how it's done in `HashOuterJoinFilteredRecordCursorFactory`